report.go

```diff
@@ -0,0 +1,358 @@
+	}
+
+	targetFile, err := os.Create(htmlPath)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := targetFile.Close(); err != nil {
+			r.logger.Error("Failed to close target file", "error",
+				err)
+		}
+	}()
+
+	return tmpl.Execute(targetFile, struct {
+		Target  string
+		History []TargetHistory
+	}{r.target, history})
+}
+
+// updateReport runs the fuzz target’s tests with coverage, generates an HTML
+// coverage report, and updates both the master index and the per-target history
+func updateReport(ctx context.Context, pkg, target string, cfg *Config,
+	logger *slog.Logger) error {
+
+	// Determine the package and corpus paths.
+	pkgPath := filepath.Join(cfg.Project.SrcDir, pkg)
+	corpusSrc := filepath.Join(cfg.Project.CorpusDir, pkg, "testdata",
+		"fuzz", target)
+	corpusDst := filepath.Join(pkgPath, "testdata", "fuzz", target)
+
+	// Copy any existing corpus files into the testdata directory.
+	if err := copyCorpusFiles(corpusSrc, corpusDst, logger); err != nil {
+		return fmt.Errorf("corpus copy failed: %w", err)
+	}
+
+	// Run `go test` for this target with coverage profiling enabled.
+	testCmd := []string{"test", fmt.Sprintf("-run=^%s$", target),
+		"-coverprofile=coverage.out", "-covermode=count"}
+	testOutput, err := runGoCommand(ctx, pkgPath, testCmd)
+	if err != nil {
+		return fmt.Errorf("go test failed for %q: %w ", pkg, err)
+	}
+
+	// Parse the coverage percentage from the test output.
+	coverageRe := regexp.MustCompile(`coverage:\s+([\d.]+)%`)
+	matches := coverageRe.FindStringSubmatch(testOutput)
+	if len(matches) < 2 {
+		return fmt.Errorf("coverage not found in output:\n%s",
+			testOutput)
+	}
+	coveragePct := matches[1]
+
+	// Generate an HTML coverage report using `go tool cover`.
+	targetReportDir := filepath.Join(cfg.Project.ReportDir, "targets",
+		pkg, target)
+	if err := EnsureDirExists(targetReportDir); err != nil {
+		return fmt.Errorf("create target report directory: %w", err)
+	}
+
+	htmlFileName := time.Now().Format("2006-01-02") + ".html"
+	reportPath := filepath.Join(targetReportDir, htmlFileName)
+
+	coverCmd := []string{"tool", "cover", "-html=coverage.out", "-o",
+		reportPath}
+	if _, err := runGoCommand(ctx, pkgPath, coverCmd); err != nil {
+		return fmt.Errorf("go tool cover failed for %q: %w ", pkg, err)
+	}
+
+	covReport := &TargetPkgReport{
+		logger:         logger,
+		pkg:            pkg,
+		target:         target,
+		coverage:       coveragePct,
+		reportDir:      cfg.Project.ReportDir,
+		reportHTMLPath: filepath.Join(target, htmlFileName),
+	}
+
+	// Record this run in the target's history and regenerate its HTML.
+	if err := covReport.updateTarget(); err != nil {
+		return fmt.Errorf("target history update failed: %w", err)
+	}
+
+	return nil
+}
+
+// copyCorpusFiles copies corpus files from source to destination directory.
+func copyCorpusFiles(srcDir, dstDir string, logger *slog.Logger) error {
+	entries, err := os.ReadDir(srcDir)
```

Comment:
Is it ever possible that the corpus `srcDir` doesn't exist?  Perhaps if the target crashes before any inputs are saved?

If so, we end up sending an error up the call stack to `runWorker`, where IIUC all the workers get cancelled.